### PR TITLE
CRM-21521: fix nested multipart handling in bounce processing

### PR DESCRIPTION
**Pull request by ejegg**
When for example a multipart-report email body's first part is
multipart-related, extract the text correctly instead of crashing.

This logic should handle all kinds of weird multipart nesting, up
to MIME_MAX_RECURSION levels deep.

Overview
----------------------------------------
Makes email bounce processing more robust

Before
----------------------------------------
Bounce processing crashes when a multipart-report email body's first part is multipart-related

After
----------------------------------------
Bounce processing correctly extracts arbitrarily nested multiparts

Technical Details
----------------------------------------
Pulls various kinds of multipart extraction into their own functions that can be called recursively. Gives up after a set recursion depth.

---

 * [CRM-21521: Bounce processing exception: no such property 'text'](https://issues.civicrm.org/jira/browse/CRM-21521)

### CODE REVIEW

File: tests/phpunit/CRM/Utils/Mail/EmailProcessorTest.php [hunk 1]
```diff
@@ -69,6 +69,21 @@ public function testProcessingMultipartRelatedEmail() {
...
+   * Sample anonymized from an email that broke bounce processing at Wikimedia
+   */
+  public function testProcessingNestedMultipartEmail() {
+    $this->setUpMailing();
+    $mail = 'test_nested_message.eml';
+
+    copy(__DIR__ . '/data/bounces/' . $mail, __DIR__ . '/data/mail/' .   $mail);
+    $this->callAPISuccess('job', 'fetch_bounces', array());
+    $this->assertFalse(file_exists(__DIR__ . '/data/mail/' . $mail));
+    $this->checkMailingBounces(1);
```
Comment (seamuslee001) [inline tests/phpunit/CRM/Utils/Mail/EmailProcessorTest.php]: @ejegg Just wondering if its worth while checking for a specific type and or message stored in the db associated with the bounce also?

**Comment by ejegg** [comment]
Any thoughts on the 'style error'? I can't figure out what it's complaining about.

File: CRM/Utils/Mail/EmailProcessor.php [hunk 2]
```diff
@@ -475,4 +428,103 @@ public static function _process($civiMail, $dao, $is_create_activities) {
...
+      return NULL;
+    }
+    $recursionLevel += 1;
+    $text = NULL;
+    if ($multipart instanceof ezcMailMultipartReport) {
+      $text = self::getTextFromMulipartReport($multipart, $recursionLevel);
+    }
+    elseif ($multipart instanceof ezcMailMultipartRelated) {
+      $text = self::getTextFromMultipartRelated($multipart, $recursionLevel);
+    } else {
```
Comment (seamuslee001) [inline CRM/Utils/Mail/EmailProcessor.php]: else needs to be on a new line

File: CRM/Utils/Mail/EmailProcessor.php [hunk 3]
```diff
@@ -475,4 +428,103 @@ public static function _process($civiMail, $dao, $is_create_activities) {
...
+      foreach ($multipart->getParts() as $part) {
+        if (isset($part->subType) and $part->subType === 'plain') {
+          $text = $part->text;
+        } elseif ($part instanceof ezcMailMultipart) {
```
Comment (seamuslee001) [inline CRM/Utils/Mail/EmailProcessor.php]: elseif needs to be on a new line

File: CRM/Utils/Mail/EmailProcessor.php [hunk 4]
```diff
@@ -475,4 +428,103 @@ public static function _process($civiMail, $dao, $is_create_activities) {
...
+   * @param int $recursionLevel
+   *
+   * @return array
+   */
+  protected static function getTextFromMultipartRelated($related, $recursionLevel) {
+    $text = NULL;
+    foreach ($related->getRelatedParts() as $part) {
+      if (isset($part->subType) and $part->subType === 'plain') {
+        $text = $part->text;
+      } elseif ($part instanceof ezcMailMultipart) {
```
Comment (seamuslee001) [inline CRM/Utils/Mail/EmailProcessor.php]: elseif needs to be on a new line

File: CRM/Utils/Mail/EmailProcessor.php [hunk 5]
```diff
@@ -475,4 +428,103 @@ public static function _process($civiMail, $dao, $is_create_activities) {
...
+    elseif (($part = $multipart->getReadablePart()) !== NULL) {
+      if (isset($part->text)) {
+        $text = $part->text;
+      }
+      elseif ($part instanceof ezcMailMultipart) {
+        $text = self::getTextFromMultipart($part, $recursionLevel);
+      }
+    }
+    return $text;
+  }
```
Comment (seamuslee001) [inline CRM/Utils/Mail/EmailProcessor.php]: need a blank line between this closing brace and the class closing brace

**Comment by seamuslee001** [comment]
@ejegg see my comments just now

File: CRM/Utils/Mail/EmailProcessor.php (hunk 5 above)
Comment (seamuslee001) [reply CRM/Utils/Mail/EmailProcessor.php]: @ejegg you need 1 complete blank line after this brace but before the } for the class

**Comment by ejegg** [comment]
Thanks seamuslee001! I can squash all those changes together to make the commit history cleaner

**Comment by ejegg** [comment]
@seamuslee001 this is all squashed and passing tests!

**Comment by seamuslee001** [comment]
nice work @ejegg :-) I'll see if i can get a colleague of mine to review it

File: CRM/Utils/Mail/EmailProcessor.php [hunk 6]
```diff
@@ -475,4 +428,107 @@ public static function _process($civiMail, $dao, $is_create_activities) {
...
+    elseif ($multipart instanceof ezcMailMultipartRelated) {
+      $text = self::getTextFromMultipartRelated($multipart, $recursionLevel);
+    }
+    else {
+      foreach ($multipart->getParts() as $part) {
+        if (isset($part->subType) and $part->subType === 'plain') {
+          $text = $part->text;
+        }
+        elseif ($part instanceof ezcMailMultipart) {
+          $text = self::getTextFromMultipart($part);
```
Comment (seamuslee001) [inline CRM/Utils/Mail/EmailProcessor.php]: Don't we need to be passing through the $recussionLevel here?

**Comment by ejegg** [comment]
Good catch @seamuslee001, fixed!

**Comment by eileenmcnaughton** [comment]
I can't merge this as @ejegg is a colleague but I merged the same patch into our wmf & want to add some notes as to the review I did there

1) I found a less complex multipart email and introduced a unit test for that sample into the main CiviCRM test suite (already merged)
2) I went through the extraction process locally, noting that the only significant changes related to the recursion issue on the email being processed
3) checked that running the unit test per this PR on the anonymised version of the email did fail just as it did on our production & that with the changes it now passes.

This code will only be reached when processing a multipart email & we have assurances from the added unit test that we have not broken that.

**Comment by colemanw** [comment]
I'm a little wary of adding global constants because of potential namespace conflicts. Any reason MAIL_MAX_RECURSION can't be added to the class instead of globally?

**Comment by ejegg** [comment]
Thanks for taking a look @colemanw! I've changed the MAX_RECURSION bit to a class constant.

**Comment by ejegg** [comment]
Looks like some unrelated DB errors in the testing setup

**Comment by ejegg** [comment]
jenkins re-test this please

**Comment by ejegg** [comment]
jenkins re-test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by colemanw** [comment]
I'm satisfied that this has had enough unit testing and code review to merge. Good work @ejegg

**Pull request merged**